core/state_prefetcher.go

```diff
@@ -59,9 +81,54 @@ func (p *statePrefetcher) Prefetch(block *types.Block, statedb *state.StateDB, c
 	)
 	workers.SetLimit(max(1, 4*runtime.NumCPU()/5)) // Aggressively run the prefetching
 
-	// Iterate over and process the individual transactions
-	for i, tx := range block.Transactions() {
+	var (
+		processed = make(map[common.Hash]struct{}, len(block.Transactions()))
+		heavyTxs  = make(chan *types.Transaction, len(block.Transactions()))
+		normalTxs = make(chan *types.Transaction, len(block.Transactions()))
+	)
+	for _, tx := range block.Transactions() {
+		// Note: gasLimit is not equivalent to gasUsed. Ideally, transaction heaviness
+		// should be measured using gasUsed. However, gasUsed is unknown prior to
+		// execution, so gasLimit is used as the indicator instead. This allows transaction
+		// senders to inflate gasLimit to gain higher prefetch priority, but this
+		// trade-off is unavoidable.
+		if isHeavyTransaction(tx.Gas(), block.GasUsed()) {
+			heavyTxs <- tx
+		}
+		// The heavy transaction will also be emitted with the normal prefetching
+		// ordering, depends on in which track it will be selected first.
+		normalTxs <- tx
+	}
+	blockPrefetchHeavyTxsMeter.Mark(int64(len(heavyTxs)))
+
+	fetchTx := func() (*types.Transaction, bool) {
+		// Pick the heavy transaction first based on the pre-defined probability
+		if rand.Intn(100) < heavyTransactionPriority {
+			select {
+			case tx := <-heavyTxs:
+				return tx, false
+			default:
+			}
+		}
+		// No more heavy transaction, or no priority for them, pick the transaction
+		// with normal order.
+		select {
+		case tx := <-normalTxs:
+			return tx, false
+		default:
+			return nil, true
+		}
+	}
+	for {
+		tx, done := fetchTx()
+		if done {
+			break
+		}
+		if _, exists := processed[tx.Hash()]; exists {
+			continue
+		}
```

Comment:
It seems like `processed` is checked but never populated. If that is correct, then transactions in `heavyTxs` will be prefetched twice.  Seems like you would either need to make `heavyTxs` and `normalTxs` mutually exclusive, or do something like this:

```go
if _, exists := processed[tx.Hash()]; exists {
	continue
}
processed[tx.Hash()] = struct{}{}
```
to ensure the transactions in `heavyTxs` are marked as processed.  Please let me know if I’m missing something.

Reply:
Good catch!